command/harness/capacity.go

```diff
@@ -0,0 +1,229 @@
+	"github.com/sirupsen/logrus"
+)
+
+type CapacityReservationRequest struct {
+	ID              string               `json:"id"` // stage runtime ID
+	PoolID          string               `json:"pool_id"`
+	FallbackPoolIDs []string             `json:"fallback_pool_ids"`
+	Tags            map[string]string    `json:"tags"`
+	CorrelationID   string               `json:"correlation_id"`
+	LogKey          string               `json:"log_key"`
+	Context         Context              `json:"context,omitempty"`
+	ResourceClass   string               `json:"resource_class"`
+	VMImageConfig   lespec.VMImageConfig `json:"vm_image_config"`
+	LogConfig       api.LogConfig        `json:"log_config"`
+	MtlsConfig      lespec.MtlsConfig    `json:"mtls_config"`
+	StorageConfig   types.StorageConfig  `json:"storage_config"`
+	Zone            string               `json:"zone"`
+	MachineType     string               `json:"machine_type"`
+	Timeout         int64                `json:"timeout,omitempty"`
+	Envs            map[string]string    `json:"envs,omitempty"`
+}
+
+// HandleCapacityReservation tries to reserve capacity for a future vm init an in any of the pools given in the
+// setup request. It calls handleCapacityReservation internally for each pool instance trying to complete a
+// capacity reservation.
+func HandleCapacityReservation(
+	ctx context.Context,
+	r *CapacityReservationRequest,
+	crs store.CapacityReservationStore,
+	globalVolumes []string,
+	poolMapByAccount map[string]map[string]string,
+	runnerName string,
+	poolManager drivers.IManager,
+	metrics *metric.Metrics,
+) (*types.CapacityReservation, error) {
+	stageRuntimeID := r.ID
+
+	if stageRuntimeID == "" {
+		return nil, ierrors.NewBadRequestError("mandatory field 'id' in the request body is empty")
+	}
+
+	if r.PoolID == "" {
+		return nil, ierrors.NewBadRequestError("mandatory field 'pool_id' in the request body is empty")
+	}
+
+	// Sets up logger to stream the logs in case log config is set
+	log := logrus.New()
+	internalLog := logrus.New()
+	internalLog.SetFormatter(&logrus.JSONFormatter{})
+
+	var (
+		logr         *logrus.Entry
+		internalLogr *logrus.Entry
+	)
+	if r.LogConfig.URL == "" {
+		log.Out = os.Stdout
+		logr = log.WithField("api", "dlite:setup").WithField("correlationID", r.CorrelationID)
+	} else {
+		wc := getStreamLogger(r.LogConfig, r.MtlsConfig, r.LogKey, r.CorrelationID)
+		defer func() {
+			if err := wc.Close(); err != nil {
+				log.WithError(err).Debugln("failed to close log stream")
+			}
+		}()
+
+		log.Out = wc
+		log.SetLevel(logrus.TraceLevel)
+		internalLog.SetLevel(logrus.TraceLevel)
+		internalLog.Out = os.Stdout
+		logr = log.WithField("stage_runtime_id", stageRuntimeID)
+		internalLogr = internalLog.WithField("stage_runtime_id", stageRuntimeID)
+	}
+
+	logr = AddContext(logr, &r.Context, r.Tags)
+	internalLogr = AddContext(internalLogr, &r.Context, r.Tags)
+	ctx = logger.WithContext(ctx, logger.Logrus(internalLogr))
```

Comment:
Just create internal logger only here. Cap reservation logs will not be streamed on UI.